### ISSUE TITLE
Pyright reports "not awaitable" even when a class has `def __await__`

### ISSUE THREAD

**Issue by myuanz** [microsoft/pyright#6844]
I'm using [aiomysql](https://github.com/aio-libs/aiomysql), I have some code that have no pyright errors in the past, but today I found pyright reports an error, my code is: 

```python
pool = await aiomysql.create_pool(
        host=config['mysql_host'],
        user=config['mysql_user'],
        password=config['mysql_password'],
        db=config['mysql_db'],
        loop=asyncio.get_running_loop(),
    )
# same to https://github.com/aio-libs/aiomysql?tab=readme-ov-file#basic-example
```

and run pyright:

```bash
$ pyright --version 
pyright 1.1.343
$ pyright web_app.py
error: “_PoolContextManager”不可等待 (reportGeneralTypeIssues)
# it means _PoolContextManager is not awaitable 
# I don't know how to configure Pyright to report errors in English
```

I had checked the source code and found it return a class named `_PoolContextManager`, this class is defined at [here](https://github.com/aio-libs/aiomysql/blob/83aa96e12b1b3f2bd373f60a9c585b6e73f40f52/aiomysql/utils.py#L30) with an `__await__` define.

Thank you for this helpful tool!

**Comment by erictraut** [comment]
Could you please provide a self-contained minimal code sample? Your sample above references symbols that are undefined.

**Comment by myuanz** [comment]
```python
# From https://github.com/aio-libs/aiomysql?tab=readme-ov-file#example-of-sqlalchemy-optional-integration
import asyncio
import aiomysql


async def test_example(loop):
    pool = await aiomysql.create_pool(host='127.0.0.1', port=3306,
                                      user='root', password='',
                                      db='mysql', loop=loop)
    async with pool.acquire() as conn:
        async with conn.cursor() as cur:
            await cur.execute("SELECT 42;")
            print(cur.description)
            (r,) = await cur.fetchone()
            assert r == 42
    pool.close()
    await pool.wait_closed()


loop = asyncio.get_event_loop()
loop.run_until_complete(test_example(loop))
```

Sorry, I forgot it, just run `pip install aiomysql` to install the package.

**Comment by erictraut** [comment]
Thanks, this will be fixed in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.344, which I just published. It will also be included in a future release of pylance.

**Comment by diego898** [comment]
Hello @erictraut - I came across this issue as well today using async redis

```python
import asyncio
from redis.asyncio import Redis
host = "localhost"

async def setup():
    redis = Redis(host=host, decode_responses=True, socket_connect_timeout=1)
    val = await redis.sadd("test", 12)
    result = await redis.smembers("test")
    print(result)

if __name__ == "__main__":
    asyncio.run(setup())
```

This works as you expect when you run it.

I'm using redis 5.0.1, and pyright 1.1.351

Pyright returns:

```
❯ poetry run pyright temp.py
temp.py
  temp.py:7:17 - error: "int" is not awaitable
    "int" is incompatible with protocol "Awaitable[_T_co@Awaitable]"
      "__await__" is not present (reportGeneralTypeIssues)
 temp.py:8:20 - error: "Set[Unknown]" is not awaitable
    "Set[Unknown]" is incompatible with protocol "Awaitable[_T_co@Awaitable]"
      "__await__" is not present (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations 
```

you can see that class has an `__await__`: https://github.com/redis/redis-py/blob/d529c2ad8d2cf4dcfb41bfd93ea68cfefd81aa66/redis/asyncio/client.py#L369

**Comment by erictraut** [comment]
@diego898, yes, pyright's behavior is correct here. The `redis.sadd` method returns a value of type `Awaitable[int] | int`. If it returns an `int`, your code will crash. Likewise, `redis.smembers` returns a value of type `Awaitable[set] | set`.

**Comment by erictraut** [comment]
Type checkers like pyright use the information supplied in type annotations in their analysis. If you suspect that the type annotations in this library are incorrect, you should report the issue to the maintainer of the library.

Based on the type information supplied, pyright is behaving correctly here. The function claims to return either an `Awaitable[int]` or an `int`, and only one of those two types is awaitable. 

I'm not familiar with this library, so I don't know how it is told whether to use async or sync mode. Depending on how this works, the library authors have a number of options for providing better type annotations that correctly and unambiguously identify the return type.

In any case, this is not a bug in pyright.

**Comment by diego898** [comment]
Thank you for explaining!

**Comment by diego898** [comment]
Thank you for explaining! This was very helpful and I found this issue:

https://github.com/redis/redis-py/issues/3091

as well
